telegram/ext/_updater.py

```diff
@@ -207,60 +179,138 @@ def start_polling(
-                    dispatcher_ready.wait()
+                self._logger.debug('Waiting for polling to start')
+                await polling_ready.wait()
+                self._logger.debug('Polling to started')
 
-                # Return the update queue so the main thread can insert updates
                 return self.update_queue
-            return None
+            except Exception as exc:
+                self._running = False
+                raise exc
+
+    async def _start_polling(
+        self,
+        poll_interval: float,
+        timeout: int,
+        read_timeout: Optional[float],
+        write_timeout: ODVInput[float],
+        connect_timeout: ODVInput[float],
+        pool_timeout: ODVInput[float],
+        bootstrap_retries: int,
+        drop_pending_updates: Optional[bool],
+        allowed_updates: Optional[List[str]],
+        ready: asyncio.Event,
+        error_callback: Optional[Callable[[TelegramError], None]],
+    ) -> None:
+
+        self._logger.debug('Updater started (polling)')
+
+        await self._bootstrap(
+            bootstrap_retries,
+            drop_pending_updates=drop_pending_updates,
+            webhook_url='',
+            allowed_updates=None,
+        )
+
+        self._logger.debug('Bootstrap done')
+
+        async def polling_action_cb() -> bool:
+            updates = await self.bot.get_updates(
+                offset=self.last_update_id,
+                timeout=timeout,
+                read_timeout=read_timeout,
+                connect_timeout=connect_timeout,
+                write_timeout=write_timeout,
+                pool_timeout=pool_timeout,
+                allowed_updates=allowed_updates,
+            )
+
+            if updates:
+                if not self.running:
+                    self._logger.critical(
+                        'Updater stopped unexpectedly. Pulled updates will be ignored and again '
+                        'on restart.'
+                    )
+                else:
+                    for update in updates:
+                        await self.update_queue.put(update)
+                    self.last_update_id = updates[-1].update_id + 1
+
+            return True
+
+        def default_error_callback(exc: TelegramError) -> None:
+            self._logger.exception('Exception happened while polling for updates.', exc_info=exc)
+
+        # Start task that runs in background, pulls
+        # updates from Telegram and inserts them in the update queue of the
+        # Application.
+        self.__polling_task = asyncio.create_task(
+            self._network_loop_retry(
+                action_cb=polling_action_cb,
+                on_err_cb=error_callback or default_error_callback,
+                description='getting Updates',
+                interval=poll_interval,
+            )
+        )
 
-    def start_webhook(
+        if ready is not None:
+            ready.set()
+
+    async def start_webhook(
```

Comment:
does start_webhook need the parameter `ready` like start_polling has? Application.run_webhook also has it.

Reply:
I'm not sure if I can follow. Neither of `Updater.start_{polling, webhook}` currently has a `ready` parameter, but both internally use such a parameter of `Updater._start_{p, w}` so that they return only when the setup is done. `Application.run_{p, w}` both a have `ready` parameter that's passed to `Application.start`.

I agree that there's an asymmetie between `App.start` and `Updater.start_*`.

The need for such a parameter was much higher in a threaded context - `Dispatcher/Updater.start()` would more or less immediately return even though the background threads were still starting up. Now even if you do `task = asyncio.create_task(App/Updater.start_*)`, you can just `await task` to ensure that the starting process is actually done. So IMO we can even remove the `ready` parameter from `Application.start`. What do you think?

Reply:
> Neither of `Updater.start_{polling, webhook}` currently has a `ready` parameter, 

Ah sorry I had seen for updater._start_polling


>  So IMO we can even remove the `ready` parameter from `Application.start`. What do you think?

Okay, sounds good 👍 can we also remove for updater._start_polling/webhook?